File: euxis-cpp/euxis-bridge-cpp/src/policy.cpp
#include <euxis/bridge/policy.hpp>

namespace euxis::bridge {

auto SkillExecutionPolicy::permissive() -> SkillExecutionPolicy {
    SkillExecutionPolicy policy;
    policy.require_signature = false;
    policy.filesystem.read_only = false;
    policy.network.deny_all = false;
    policy.resources.memory_mb = 512;
    policy.resources.timeout_seconds = 60;
    return policy;
}

auto SkillExecutionPolicy::to_json() const -> nlohmann::json {
    nlohmann::json j;
    j["require_signature"] = require_signature;
    if (public_key_path) {
        j["public_key_path"] = public_key_path->string();
    }
    j["filesystem"]["read_only"] = filesystem.read_only;
    nlohmann::json write_paths = nlohmann::json::array();
    for (const auto& p : filesystem.write_paths) {
        write_paths.push_back(p.string());
    }
    j["filesystem"]["write_paths"] = write_paths;
    j["network"]["deny_all"] = network.deny_all;
    j["network"]["allowed_hosts"] = network.allowed_hosts;
    j["resources"]["memory_mb"] = resources.memory_mb;
    j["resources"]["timeout_seconds"] = resources.timeout_seconds;
    if (output_schema) {
        j["output_schema"] = *output_schema;
    }
    if (audit_log_path) {
        j["audit_log_path"] = audit_log_path->string();
    }
    return j;
}

}  // namespace euxis::bridge

File: euxis-bridge/src/euxis_bridge/core/policy.py
"""Execution policy for bridged skill sandboxing and verification."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True, slots=True)
class FilesystemPolicy:
    """Controls filesystem access for skill execution."""

    read_only: bool = True
    write_paths: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class NetworkPolicy:
    """Controls network access for skill execution."""

    deny_all: bool = True
    allowed_hosts: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class ResourceLimits:
    """Resource bounds for skill execution."""

    memory_mb: int = 256
    timeout_seconds: int = 20


@dataclass(frozen=True, slots=True)
class SkillExecutionPolicy:
    """Policy governing bridged skill execution."""

    require_signature: bool = False
    public_key_path: str | None = None
    filesystem: FilesystemPolicy = field(default_factory=FilesystemPolicy)
    network: NetworkPolicy = field(default_factory=NetworkPolicy)
    resources: ResourceLimits = field(default_factory=ResourceLimits)
    output_schema: dict[str, Any] | None = None
    audit_log_path: str | None = None

    @classmethod
    def permissive(cls) -> SkillExecutionPolicy:
        """Create a permissive policy suitable for trusted local skills."""
        return cls(
            require_signature=False,
            filesystem=FilesystemPolicy(read_only=False),
            network=NetworkPolicy(deny_all=False),
            resources=ResourceLimits(memory_mb=512, timeout_seconds=60),
        )
